rust/src/transformations/make_stable_expr/expr_discrete_quantile_score/plugin_dq_score.rs

```diff
@@ -20,25 +22,104 @@ use polars_plan::{
+pub(crate) struct DiscreteQuantileScoreArgs {
     /// Candidates to score
-    pub candidates: Vec<f64>,
+    pub candidates: Candidates,
     /// A value between [0, 1]
     pub alpha: f64,
     /// Alpha numerator, alpha denominator, and max partition length
     pub constants: Option<(u64, u64, u64)>,
 }
 
-impl OpenDPPlugin for DQScoreArgs {
+#[derive(Clone)]
+pub(crate) struct Candidates(pub Series);
+
+#[cfg(feature = "ffi")]
+impl Serialize for Candidates {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(self.0.len()))?;
+
+        for value in self.0.iter() {
+            match value {
+                AnyValue::Boolean(v) => seq.serialize_element(&v),
+                AnyValue::Int64(v) => seq.serialize_element(&v),
+                AnyValue::Float64(v) => seq.serialize_element(&v),
+                AnyValue::String(v) => seq.serialize_element(&v),
+                _ => Err(serde::ser::Error::custom(
+                    "Expected homogenous candidates of either bool, i64, f64, or string",
+                )),
+            }?;
+        }
+        seq.end()
+    }
+}
+
+#[cfg(feature = "ffi")]
+impl<'de> Deserialize<'de> for Candidates {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let value = Value::deserialize(deserializer)?;
+
+        let Value::List(list) = value else {
+            return Err(serde::de::Error::custom(format!(
+                "Expected a list, found {:?}",
+                value
+            )));
+        };
+        let first = list
+            .first()
+            .ok_or_else(|| serde::de::Error::custom("Expected at least one candidate"))?
+            .clone();
+
+        macro_rules! match_candidates {
+            ($list:ident, $dtype:ident) => {
+                Series::new(
+                    "",
+                    $list
+                        .into_iter()
+                        .map(|x| match x {
+                            Value::$dtype(v) => Ok(v),
+                            _ => Err(serde::de::Error::custom("Expected homogenous candidates")),
+                        })
+                        .collect::<Result<Vec<_>, D::Error>>()?,
+                )
+            };
+        }
+
+        let candidates = match first {
+            Value::Bool(_) => match_candidates!(list, Bool),
+            Value::I64(_) => match_candidates!(list, I64),
+            Value::F64(_) => match_candidates!(list, F64),
+            Value::String(_) => match_candidates!(list, String),
+            first => {
+                return Err(serde::de::Error::custom(format!(
+                    "Candidates must be homogeneous primitives, first candidate is {:?}",
+                    first
+                )))
+            }
+        };
+
+        Ok(Candidates(candidates))
+    }
+}
```

Comment:
A test of serialization and deserialization would be easy to add.

Reply:
It gets moved in https://github.com/opendp/opendp/pull/1498/ and tested there. Mind if we wait?